Tests/GRPCTests/GRPCReflectionServiceTests/GRPCReflectionServiceTests.swift

```diff
@@ -0,0 +1,362 @@
+          XCTAssert(protos[protoIndex!].messageType.contains(messageType))
+        }
+      }
+      protos.removeAll { $0 == fileDescriptorProto }
+    }
+    XCTAssert(protos.isEmpty)
+  }
+
+  func testGetSerializedFileDescriptorProtosDependencyLoops() throws {
+    var protos = self.makeProtosWithDependencies()
+    // Making dependencies of the "bar1.proto" to depend on "bar1.proto".
+    protos[1].dependency.append("bar1.proto")
+    protos[2].dependency.append("bar1.proto")
+    protos[3].dependency.append("bar1.proto")
+    let registry = try ReflectionServiceData(fileDescriptor: protos)
+    let serializedFileDescriptorProtos =
+      try registry
+      .getSerializedFileDescriptorProtos(fileName: "bar1.proto")
+    let fileDescriptorProtos = try serializedFileDescriptorProtos.map {
+      try Google_Protobuf_FileDescriptorProto(serializedData: $0)
+    }
+    // Test that we get only 4 serialized File Descriptor Protos as response.
+    XCTAssertEqual(fileDescriptorProtos.count, 4)
+    for fileDescriptorProto in fileDescriptorProtos {
+      XCTAssert(protos.contains(fileDescriptorProto))
+      let protoIndex = protos.firstIndex(of: fileDescriptorProto)
+      for service in fileDescriptorProto.service {
+        XCTAssert(protos[protoIndex!].service.contains(service))
+        let serviceIndex = protos[protoIndex!].service.firstIndex(of: service)
+        let originalMethods = protos[protoIndex!].service[serviceIndex!].method
+        for method in service.method {
+          XCTAssert(originalMethods.contains(method))
+        }
+        for messageType in fileDescriptorProto.messageType {
+          XCTAssert(protos[protoIndex!].messageType.contains(messageType))
+        }
+      }
+      protos.removeAll { $0 == fileDescriptorProto }
+    }
+    XCTAssert(protos.isEmpty)
+  }
+
+  func testGetSerializedFileDescriptorProtosInvalidFile() throws {
+    let protos = self.makeProtosWithDependencies()
+    let registry = try ReflectionServiceData(fileDescriptor: protos)
+    XCTAssertThrowsError(
+      try registry.getSerializedFileDescriptorProtos(fileName: "invalid.proto")
+    ) { error in
+      XCTAssertEqual(
+        error as! GRPCStatus,
+        GRPCStatus(
+          code: .notFound,
+          message: "The provided file or a dependency of the provided file could not be found."
+        )
+      )
+    }
+  }
+
+  func testGetSerializedFileDescriptorProtosDependencyNotProto() throws {
+    var protos = self.makeProtosWithDependencies()
+    protos[0].dependency.append("invalidDependency")
+    let registry = try ReflectionServiceData(fileDescriptor: protos)
+    XCTAssertThrowsError(
+      try registry.getSerializedFileDescriptorProtos(fileName: "bar1.proto")
+    ) { error in
+      XCTAssertEqual(
+        error as! GRPCStatus,
+        GRPCStatus(
+          code: .notFound,
+          message: "The provided file or a dependency of the provided file could not be found."
+        )
+      )
+    }
```

Comment:
There's an interesting edge case here that you'll need to look into. I'm actually not sure if this is a valid test (it might be).

There are some protobuf messages whose generated code is distributed with the language they are for. For example SwiftProtobuf comes with a copy of `Google_Protobuf_Any` (https://github.com/apple/swift-protobuf/blob/main/Sources/SwiftProtobuf/Google_Protobuf_Any%2BExtensions.swift). If you depend on that type in the definition of another proto message then you don't need to provide the `.proto` file when generating Swift code (I believe this is true for other languages too).

So I don't know if the transitive dependencies need to include these "standard" messages or not. I would start by looking at the file descriptor proto for a message uses e.g. `Google_Protobuf_Any` and seeing what the dependencies are.